drivers/serial/uart_emul.c

```diff
@@ -99,15 +112,105 @@ int uart_emul_config_get(const struct device *dev, struct uart_config *cfg)
 }
 #endif /* CONFIG_UART_USE_RUNTIME_CONFIGURE */
 
+static int uart_emul_fifo_read(const struct device *dev, uint8_t *rx_data, int size)
+{
+	int ret;
+	struct uart_emul_data *data = dev->data;
+
+	K_SPINLOCK(&data->rx_lock)
+	{
+		ret = MIN(size, ring_buf_size_get(data->rx_rb));
+		size = ret;
+
+		for (int n = 0; size > 0; size -= n, rx_data += n) {
+			n = ring_buf_get(data->rx_rb, rx_data, size);
+		}
+	}
+
+	return ret;
+}
+
+static int uart_emul_irq_rx_ready(const struct device *dev)
+{
+	struct uart_emul_data *data = dev->data;
+
+	if (!data->irq_en) {
+		return 0;
+	}
+
+	return 1;
+}
+
+static void uart_emul_irq_handler(struct k_work *work)
+{
+	struct uart_emul_work *uwork = CONTAINER_OF(work, struct uart_emul_work, work);
+	const struct device *dev = uwork->dev;
+	struct uart_emul_data *data = dev->data;
+	void *udata = data->irq_cb_user_data;
+
+	if (!data->irq_en) {
+		return;
+	}
+
+	while (!ring_buf_is_empty(data->rx_rb)) {
+		data->irq_cb(dev, udata);
+	}
+}
+
+static int uart_emul_irq_is_pending(const struct device *dev)
+{
+	struct uart_emul_data *const data = dev->data;
+
+	return !ring_buf_is_empty(data->rx_rb);
+}
+
+static void uart_emul_irq_rx_enable(const struct device *dev)
+{
+	struct uart_emul_data *const data = dev->data;
+
+	data->irq_en = true;
+	if (!ring_buf_is_empty(data->rx_rb)) {
+		(void)k_work_submit(&data->irq_work.work);
+	}
+}
+
+static void uart_emul_irq_rx_disable(const struct device *dev)
+{
+	struct uart_emul_data *const data = dev->data;
+
+	data->irq_en = false;
+}
+
+static void uart_emul_irq_callback_set(const struct device *dev, uart_irq_callback_user_data_t cb,
+				       void *user_data)
+{
+	struct uart_emul_data *const data = dev->data;
+
+	data->irq_cb = cb;
+	data->irq_cb_user_data = user_data;
+}
+
+static int uart_emul_irq_update(const struct device *dev)
+{
+	return 1;
+}
+
 static const struct uart_driver_api uart_emul_api = {
 	.poll_in = uart_emul_poll_in,
 	.poll_out = uart_emul_poll_out,
 #ifdef CONFIG_UART_USE_RUNTIME_CONFIGURE
 	.config_get = uart_emul_config_get,
 	.configure = uart_emul_configure,
 #endif /* CONFIG_UART_USE_RUNTIME_CONFIGURE */
-	.err_check = uart_emul_err_check
-};
+	.err_check = uart_emul_err_check,
+
+	COND_CODE_1(
+		IS_ENABLED(CONFIG_UART_INTERRUPT_DRIVEN),
```

Comment:
`IF_ENABLED` would simplify this

Reply:
moved to and fixed in https://github.com/zephyrproject-rtos/zephyr/pull/63416